File: app/src/main/java/com/dcrandroid/activities/ConfirmSeedActivity.java
package com.dcrandroid.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.dcrandroid.R;
import com.dcrandroid.data.Constants;
import com.dcrandroid.util.DcrConstants;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by Macsleven on 25/12/2017.
 */

public class ConfirmSeedActivity extends AppCompatActivity {
    private List<String> seeds;
    private Set<String> tempSeeds;
    private TextView confirmView;
    private AutoCompleteTextView seedTv;
    private String seed = "";
    private ArrayAdapter<String> adapter;
    private Boolean restore;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
        setContentView(R.layout.confirm_seed_page);
        ImageView buttonSeedClear = findViewById(R.id.button_clear_seed);
        Button btnConfirmSeed = findViewById(R.id.button_confirm_seed);
        Button btnDeleteSeed = findViewById(R.id.button_delete_seed);
        seedTv = findViewById(R.id.autoCompleteSeed);
        seedTv.setSingleLine(true);
        seedTv.setCompletionHint(getString(R.string.tap_to_select));

        btnConfirmSeed.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                String enteredSeed = "";
                Intent i = new Intent(ConfirmSeedActivity.this, EncryptWallet.class)
                        .putExtra(Constants.SEED, enteredSeed);
                startActivity(i);
                return true;
            }
        });
        btnConfirmSeed.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String enteredSeed = confirmView.getText().toString().trim();
                if (!restore) {
                    if (seed.equals(enteredSeed)) {
                        Intent i = new Intent(ConfirmSeedActivity.this, EncryptWallet.class)
                                .putExtra(Constants.SEED, enteredSeed);
                        startActivity(i);
                    } else {
                        Toast.makeText(ConfirmSeedActivity.this, R.string.incorrect_seed_input, Toast.LENGTH_LONG).show();
                    }
                } else {
                    DcrConstants constants = DcrConstants.getInstance();
                    if (constants.wallet.verifySeed(enteredSeed)) {
                        Intent i = new Intent(ConfirmSeedActivity.this, EncryptWallet.class)
                                .putExtra(Constants.SEED, enteredSeed);
                        startActivity(i);
                    } else {
                        Toast.makeText(ConfirmSeedActivity.this, R.string.incorrect_seed_input, Toast.LENGTH_LONG).show();
                    }
                }
            }
        });
        buttonSeedClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                confirmView.setText("");
            }
        });
        btnDeleteSeed.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String[] enteredSeed = confirmView.getText().toString().trim().split(" ");
                ArrayList<String> temp = new ArrayList<>();
                temp.addAll(Arrays.asList(enteredSeed));
                if(temp.size() > 0){
                    //remove last seed
                    temp.remove(temp.size() - 1);
                    StringBuilder sb = new StringBuilder();
                    for(int i = 0; i < temp.size(); i++){
                        sb.append(" ");
                        sb.append(temp.get(i));
                    }
                    confirmView.setText(sb.toString().trim());
                }else{
                    confirmView.setText("");
                }
            }
        });
        confirmView = findViewById(R.id.seed_display_confirm);
        seedTv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                String s = (String)adapterView.getItemAtPosition(i);
                confirmView.setText(String.format("%s %s", confirmView.getText().toString().trim(), s));
                seedTv.setText("");
            }
        });

        prepareData();
    }

    private void prepareData() {
        Intent i = getIntent();
        Bundle b = i.getExtras();
        if(b != null) {
            seed = b.getString(Constants.SEED).trim();
            restore = b.getBoolean(Constants.RESTORE);
            seeds = new ArrayList<>();
            String[] seedsArray = seed.split(" ");
            tempSeeds = new HashSet<>(Arrays.asList(seedsArray));
            List<String> list = new ArrayList<>(tempSeeds);
            seeds.addAll(Arrays.asList(seedsArray));
            if(restore){
                Collections.sort(seeds, new SortIgnoreCase());
            }else{
                Collections.shuffle(seeds);
            }
            adapter = new ArrayAdapter<>(this,
                    android.R.layout.simple_dropdown_item_1line, list);
            seedTv.setAdapter(adapter);
        }else{
            Toast.makeText(this, R.string.error_bundle_null, Toast.LENGTH_SHORT).show();
        }
    }

    public class SortIgnoreCase implements Comparator<String>{

        @Override
        public int compare(String s, String t1) {
            return s.toLowerCase().compareTo(t1.toLowerCase());
        }
    }
}

File: app/src/main/java/com/dcrandroid/activities/ConfirmSeedActivity.kt
/*
 * Copyright (c) 2018-2019 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.activities

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.dcrandroid.R
import com.dcrandroid.adapter.InputSeed
import com.dcrandroid.adapter.RestoreWalletAdapter
import com.dcrandroid.data.Constants
import com.dcrandroid.util.PreferenceUtil
import dcrlibwallet.Dcrlibwallet
import kotlinx.android.synthetic.main.confirm_seed_page.*

private const val CLICK_THRESHOLD = 300 //millisecond

class ConfirmSeedActivity : BaseActivity(), View.OnTouchListener {

    private var verifiedSeed: Boolean = false
    private var lastConfirmClick: Long = 0
    private var finalSeedsString = ""

    private var allSeeds = ArrayList<String>()
    private val seedsForInput = ArrayList<InputSeed>()
    private val confirmedSeedsArray = ArrayList<InputSeed>()
    private var sortedList = listOf<InputSeed>()

    private var handler: Handler? = null

    private lateinit var restoreWalletAdapter: RestoreWalletAdapter
    private lateinit var linearLayoutManager: LinearLayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE)

        setContentView(R.layout.confirm_seed_page)
        recyclerViewSeeds.isNestedScrollingEnabled = false
        linearLayoutManager = LinearLayoutManager(this)
        recyclerViewSeeds.layoutManager = linearLayoutManager
        button_confirm_seed.setOnTouchListener(this)
        prepareData()
    }

    private fun prepareData() {
        val bundle = intent.extras
        val temp = arrayOfNulls<String>(33)
        if (bundle != null && !bundle.isEmpty) {
            val seed = bundle.getString(Constants.SEED)
            allSeeds = ArrayList(seed.split(" "))
            temp.forEachIndexed { number, _ ->
                seedsForInput.add(InputSeed(number, " "))
                initSeedAdapter()
            }
        }
    }

    private fun initSeedAdapter() {
        restoreWalletAdapter = RestoreWalletAdapter(seedsForInput, allSeeds, applicationContext,
                { savedSeed: InputSeed ->
                    confirmedSeedsArray.add(savedSeed)
                    sortedList = confirmedSeedsArray.sortedWith(compareBy { it.number }).distinct()
                    if (sortedList.size < 33) {
                        tvError.visibility = View.VISIBLE
                        tvError.text = getString(R.string.notAllSeedsEntered)
                    }
                    val itemView = recyclerViewSeeds.findViewById<RelativeLayout>(R.id.restoreSeedRow)
                    val itemHeight = itemView.measuredHeight + itemView.measuredHeight / 60
                    val maxAllowedHeight = nestedScrollView.getChildAt(0).bottom - itemHeight * 3
                    val currentHeight = (nestedScrollView.scrollY + nestedScrollView.height)

                    if (currentHeight > maxAllowedHeight) {
                        recyclerViewSeeds.isFocusableInTouchMode = false
                        llButtons.isFocusableInTouchMode = true
                        llButtons.requestFocus()
                    }
                },
                { removeSeed: InputSeed ->
                    confirmedSeedsArray.clear()
                    confirmedSeedsArray.addAll(sortedList)
                    confirmedSeedsArray.remove(removeSeed)
                    sortedList = confirmedSeedsArray.sortedWith(compareBy { it.number })
                }, { isAllEntered: Boolean ->
            if (isAllEntered && sortedList.size == 33) {
                handleSingleTap()
                val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(llButtons.windowToken, 0)
            }
        })
        recyclerViewSeeds.adapter = restoreWalletAdapter
    }

    override fun onTouch(v: View?, event: MotionEvent?): Boolean {
        when (event!!.action) {
            MotionEvent.ACTION_DOWN -> {
                lastConfirmClick = System.currentTimeMillis()
                handler = Handler()
                handler!!.postDelayed(longHold, 5000)
            }
            MotionEvent.ACTION_UP -> {
                handler!!.removeCallbacks(longHold)
                if ((System.currentTimeMillis() - lastConfirmClick) <= CLICK_THRESHOLD) {
                    if (verifiedSeed) {

                        val util = PreferenceUtil(this)
                        util.set(Constants.SEED, null)
                        util.setBoolean(Constants.VERIFIED_SEED, true)

                        val intent = Intent(this, EncryptWallet::class.java)
                        intent.putExtra(Constants.SEED, finalSeedsString)
                        startActivity(intent)
                    }
                }
            }
        }

        return false
    }

    private val longHold = Runnable {
        val enteredSeed = "miser stupendous backward inception slowdown Capricorn uncut visitor slowdown caravan blockade hemisphere repay article necklace hazardous cobra inferno python suspicious minnow Norwegian chairlift backwater surmount impetus cement stupendous snowslide sympathy fallout embezzle afflict"
        if (enteredSeed.isNotEmpty()) {

            val util = PreferenceUtil(this)
            util.set(Constants.SEED, null)
            util.setBoolean(Constants.VERIFIED_SEED, true)

            val i = Intent(this@ConfirmSeedActivity, EncryptWallet::class.java)
                    .putExtra(Constants.SEED, enteredSeed)
            startActivity(i)
        }
    }

    private fun handleSingleTap() {

        if (sortedList.isNotEmpty() && (sortedList.size == 33)) {
            finalSeedsString = sortedList.joinToString(" ", "", "", -1, "...") { it.phrase }
            verifiedSeed = Dcrlibwallet.verifySeed(finalSeedsString)

            if (verifiedSeed) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    button_confirm_seed.background = ContextCompat.getDrawable(applicationContext, R.drawable.btn_shape3)
                } else {
                    button_confirm_seed.setBackgroundDrawable(ContextCompat.getDrawable(applicationContext, R.drawable.btn_shape3))
                }

                tvError.visibility = View.GONE

            } else {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    button_confirm_seed.background = ContextCompat.getDrawable(applicationContext, R.drawable.btn_shape2)
                } else {
                    button_confirm_seed.setBackgroundDrawable(ContextCompat.getDrawable(applicationContext, R.drawable.btn_shape2))
                }

                tvError.visibility = View.VISIBLE
                tvError.text = getString(R.string.restore_wallet_incorrect_seed_input)
            }

        } else if (sortedList.isNotEmpty() && (sortedList.size != 33)) {
            tvError.text = getString(R.string.notAllSeedsEntered)
        } else {
            tvError.text = getString(R.string.theInputFieldIsEmpty)
        }
    }
}
